nixos/modules/services/networking/nats.nix

```diff
@@ -0,0 +1,174 @@
+        type = types.int;
+        description = ''
+          Port on which to listen.
+        '';
+      };
+
+      dataDir = mkOption {
+        default = "/var/lib/nats";
+        type = types.path;
+        description = ''
+          The data directory.
+        '';
+      };
+
+      settings = mkOption {
+        default = { };
+        type = types.submodule {
+          freeformType = with lib.types; attrsOf anything;
+          options = { };
+        };
+        example = literalExample ''
+          {
+            jetstream = {
+              max_mem = "1G";
+              max_file = "10G";
+            };
+          };
+        '';
+        description = ''
+          Declarative Nats configuration. See https://docs.nats.io/nats-server/configuration
+        '';
+      };
+    };
+  };
+
+  ### Implementation
+
+  config = mkIf cfg.enable {
+    services.nats.settings = {
+      server_name = mkDefault cfg.serverName;
+      host = mkDefault cfg.host;
+      port = mkDefault cfg.port;
+      jetstream =
+        optionalAttrs cfg.jetstream { store_dir = mkDefault cfg.dataDir; };
+    };
+
+    systemd.services.nats = {
+      description = "NATS messaging system";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      restartTriggers = [ confFile ];
+
+      serviceConfig = {
+        Type = "simple";
+        ExecStart = "${cfg.package}/bin/nats-server -c ${confFile}";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        ExecStop = "${pkgs.coreutils}/bin/kill -SIGINT $MAINPID";
+        Restart = "on-failure";
+
+        User = cfg.user;
+        Group = cfg.group;
+
+        RuntimeDirectory = "nats";
+
+        # Hardening
+        CapabilityBoundingSet = "";
+        DevicePolicy = "closed";
+        LockPersonality = true;
+        MemoryDenyWriteExecute = true;
+        NoNewPrivileges = true;
+        PrivateDevices = true;
+        PrivateTmp = true;
+        PrivateUsers = true;
+        ProtectClock = true;
+        ProtectControlGroups = true;
+        ProtectHome = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        ProtectProc = "invisible";
+        ProcSubset = "pid";
+        ProtectSystem = "strict";
+        ReadWritePaths = [ cfg.dataDir ];
+        ReadOnlyPaths = [ ];
+        RemoveIPC = true;
+        RestrictAddressFamilies = [ "AF_INET" "AF_INET6" ];
+        RestrictNamespaces = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        SystemCallArchitectures = "native";
+        SystemCallFilter = [ "@system-service" "~@privileged" "~@resources" ];
+        UMask = "0077";
+      };
+    };
+
+    users.users = mkIf (cfg.user == "nats") {
+      nats = {
+        description = "NATS daemon user";
+        isSystemUser = true;
+        group = cfg.group;
+        home = cfg.dataDir;
+        createHome = true;
```

Comment:
Is there a recommended alternative for `createHome = true` but then at service activation? With `createHome = false` the test fails because of: ```server # [    9.079281] systemd[893]: nats.service: Failed to set up mount namespacing: /run/systemd/unit-root/var/lib/nats: No such file or directory```

Reply:
You still need to create the directory. I generally recommend `StateDirectory` to provision these sorts of directories. Check out this:

https://github.com/NixOS/nixpkgs/blob/b72bde7c4a1f9c9bf1a161f0c267186ce3c6483c/nixos/modules/services/misc/gitolite.nix#L26-L35

https://github.com/NixOS/nixpkgs/blob/b72bde7c4a1f9c9bf1a161f0c267186ce3c6483c/nixos/modules/services/misc/gitolite.nix#L165-L177

How does that look to you?

Reply:
Cool, that works!